### PR TITLE
Release v3.2.1

### LINKED ISSUE #626: Manta v3.2.1 Release checklist
Closes https://github.com/Manta-Network/Manta/milestone/9

# Release Checklist

Most of the following checks should be completed before officially publishing the new release
of the Calamari/Manta runtime or client. Some need to be completed after the new code is deployed.

### Runtime Releases

These checks should be performed on the codebase prior to freezing our release candidate:

- [x] Verify [`spec_version`](#spec-version) has been incremented since the
    last release for any native runtimes from any existing use on public
    (non-private/test) networks. If the runtime was published (release or pre-release), either
    the `spec_version` or `impl` must be bumped.
- [x] Verify pallet and [extrinsic ordering](#extrinsic-ordering) has stayed
    the same. Bump `transaction_version` if not.
- [x] Verify new extrinsics have been correctly whitelisted/blacklisted
- [x] Verify [benchmarks](#benchmarks) have been updated for any modified
    runtime logic.
- [x] Check for any upstream storage migrations and perform tests with `try-runtime`, if any.

The following checks can be performed after we have frozen our release candidate:

- [x] Code freeze should typically happen one week prior to release, to ensure we have enough time for related testing.
- [x] Notify everyone, especially people with merge rights to `manta` (stechu, Dengjianping) that a release is ongoing and no more merges to `manta` should happen until told otherwise
- [x] Complete the following [manual QA workflow](https://hackmd.io/TbFmorG2RnOPmLuFcg9JOQ?view).
- [x] Verify Polkadot JS API are up to date with the latest
    runtime changes.
- [x] Execute runtime upgrade to Baikal and verify network stability.
- [x] Execute runtime upgrade to Rococo and verify network stability.
- [x] Prepare a governance post and submit to our forum with description and motivation for changes.

### Client Releases

- [x] Verify that each crate's `version` has been bumped from previous release.
- [x] Check that the new client versions have [burned-in](#burn-in) without issue for at least 12 hours.

### All Releases

- [x] Check that a draft release has been created at
    https://github.com/Manta-Network/Manta/releases with relevant [release
    notes](#release-notes)
- [x] Check that build artifacts have been added to the
    draft-release
- [x] Coordinate with marketing team for documentation updates and other relevant tasks.
- [x] Update changelog.
- [x] If the release contains any changes that break/change functionality used in https://github.com/Manta-Network/sdk (e.g. RPC changes, see also [extrinsic ordering](#extrinsic-ordering)), raise a PR there and **block this release** until your PR has been merged and incorporated in a new SDK release.

Note: Do not publish draft releases from PR branches, because those branches will be deleted when the PR is merged.

### After Runtime Upgrade
- [x] Notify subscan team. Ensure subscan service can continue to scan calamari blocks.

### Burn In

Ensure that Manta DevOps has run the new release on Como and Baikal nodes
for at least 12 hours prior to publishing the release.

### Release notes

The release notes **MUST** contain:

- The priority of the release (i.e., how quickly users should upgrade) - this is
    based on the max priority of any *client* changes.
- The version of [Manta SDK](https://github.com/Manta-Network/sdk) that is compatible with this release
- Which native runtimes and their versions are included
- The proposal hashes of the runtimes as built with [srtool](https://gitlab.com/chevdor/srtool)
- (After auditing starts:) Any changes in this release that are still awaiting audit

The release notes **MAY** also list:

- Free text at the beginning of the notes mentioning anything important
    regarding this release
- Notable changes (those labelled with B[1-9]-* labels) separated into sections

### Spec Version

A runtime upgrade must bump the spec number. This may follow a pattern with the
client release

### Extrinsic Ordering

Offline signing libraries depend on a consistent ordering of call indices and
functions. Compare the metadata of the current and new runtimes and ensure that
the `module index, call index` tuples map to the same set of functions. To generate a diff report you can do the following:
* Go to the [metadata diff tool page](https://github.com/Manta-Network/Manta/actions/workflows/metadata_diff.yml).
* Open `Run workflow` drop-down menu.
* Choose the branch you want to test and other inputs if the defaults are not applicable.
* Run the workflow.
* Wait for the job to complete and inspect the output. The things to look for in the `output.txt` are lines like:
  - `[Identity] idx 28 -> 25 (calls 15)` - indicates the index for `Identity` has changed
  - `[+] Society, Recovery` - indicates the new version includes 2 additional modules/pallets.
  - If no indices have changed, every modules line should look something like `[Identity] idx 25 (calls 15)`

 In case of a breaking change, bump the `transaction_version`.

Note: Adding new functions to the runtime does not constitute a breaking change
as long as the indexes did not change.

### Benchmarks

There are three benchmarking machines reserved for updating the weights at
release-time. To initialize a benchmark run for each production runtime
(calamari, manta):
* Go to [Calamari Benchmarking Github Action](https://github.com/Manta-Network/Manta/actions/workflows/generate_calamari_weights_files.yml) 
  and [Manta Benchmarking Github Action](https://github.com/Manta-Network/Manta/actions/workflows/generate_manta_weights_files.yml)
* Open `Run workflow` drop-down menu.
* Choose your branch and run the workflow.
* When these jobs have completed (it takes a few hours), custom weights files will
    be available to download as artifacts. 
* Commit the changes to your branch and push to the remote branch for review.
* The weights should be (Currently manually) checked to make sure there are no
    big outliers (i.e., twice or half the weight).

### Security Audit

Before release, run a `Security Audit`

* Go to [Security Audit](https://github.com/Manta-Network/Manta/actions/workflows/audit.yml).
* Open `Run workflow` drop-down menu.
* Choose your branch and run the workflow.
* An audit report will be generated.
* Address any reported findings before release. If it cannot be fixed, please file an issue.

### PR DESCRIPTION
**Pull request by Garandor**
Signed-off-by: Adam Reif <Garandor@manta.network>

## Description

relates to #626

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests.
- [x] Updated relevant documentation in the code.
- [x] Added **one** line describing your change in `<branch>/CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [x] If runtime changes, need to update the version numbers properly:
   * `authoring_version`: The version of the authorship interface. An authoring node will not attempt to author blocks unless this is equal to its native runtime.
   * `spec_version`: The version of the runtime specification. A full node will not attempt to use its native runtime in substitute for the on-chain Wasm runtime unless all of spec_name, spec_version, and authoring_version are the same between Wasm and native.
   * `impl_version`: The version of the implementation of the specification. Nodes are free to ignore this; it serves only as an indication that the code is different; as long as the other two versions are the same then while the actual code may be different, it is nonetheless required to do the same thing. Non-consensus-breaking optimizations are about the only changes that could be made which would result in only the impl_version changing.
   * `transaction_version`: The version of the extrinsics interface. This number must be updated in the following circumstances: extrinsic parameters (number, order, or types) have been changed; extrinsics or pallets have been removed; or the pallet order in the construct_runtime! macro or extrinsic order in a pallet has been changed. You can run the `metadata_diff.yml` workflow for help. If this number is updated, then the `spec_version` must also be updated 
- [x] Verify benchmarks & weights have been updated for any modified runtime logics
- [x] If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- [x] If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- [x] If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,31 +1,31 @@
...
-## [Unreleased]
-
+## v3.2.1
 ### Breaking changes
-- [\#628](https://github.com/Manta-Network/Manta/pull/628) Improve RPC performance, add `max_receivers` and `max_senders` fields in the RPC request.
+- [Dolphin] [\#628](https://github.com/Manta-Network/Manta/pull/628) Improve RPC performance, add `max_receivers` and `max_senders` fields in the RPC request.
 
 ### Features
-- [\#576](https://github.com/Manta-Network/Manta/pull/576) Unfilter xtokens.transfer_multicurrencies and bump MaxAssetsForTransfer to 2.
+- [\#576](https://github.com/Manta-Network/Manta/pull/576) Unfilter `xtokens.transfer_multicurrencies` and bump MaxAssetsForTransfer to 2.
```
Comment (Dengjianping) [inline CHANGELOG.md]: Not sure why this change occur in both sections `Breaking changes` and `Improvements`.

Comment (Garandor) [reply CHANGELOG.md]: Breaking changes and improvements would be okay.
Breaking is a kind of spotlight section with changes that need extra attention and any kind of change bugfix/imprvement/new feature can break something existing

what shouldnt be the case is that 576 is both a feature and an improvement.

Comment (Garandor) [reply CHANGELOG.md]: Fixed it

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,32 +1,31 @@
...
 - [\#646](https://github.com/Manta-Network/Manta/pull/646) Add collator session keys for future nimbus consensus and a vrf placeholder
 
 ### Improvements
 - [\#449](https://github.com/Manta-Network/Manta/pull/449) Remove strip from CI, and add strip profile to production.
 - [\#571](https://github.com/Manta-Network/Manta/pull/571) Update upstream dependencies to v0.9.22.
 - [\#563](https://github.com/Manta-Network/Manta/pull/563) Re-implement the `TransactAsset` trait with the unified interface of `FungibleLedger` trait, and `AssetConfig` trait.
 - [\#576](https://github.com/Manta-Network/Manta/pull/576) Unfilter xtokens.transfer_multicurrencies and bump MaxAssetsForTransfer to 2.
-- [\#607](https://github.com/Manta-Network/Manta/pull/607) Turn node client code into library for CLI project
+- [\#607](https://github.com/Manta-Network/Manta/pull/607) Turn node client code into library for CLI project.
+- [Dolphin] [\#609](https://github.com/Manta-Network/Manta/pull/609) Update parameter path from `sdk` to `manta-parameters`.
```
Comment (Garandor) [inline CHANGELOG.md]: 609 was closed, not merged. it's ideantical to 622 which is in here already

**Comment by Dengjianping** [comment]
There're two new extrinsics added, comparing to last release(v3.2.0).
```
force_batch => pallet-utility: root origin
remove_approval => pallet-treasury: EnsureRootOrMoreThanHalfCouncil
```

But both are unnecessary to be filtered.

**Review by ghzlatarev — DISMISSED** [review]

File: runtime/calamari/src/weights/cumulus_pallet_xcmp_queue.rs [hunk 3]
```diff
@@ -53,13 +54,13 @@ pub struct SubstrateWeight<T>(PhantomData<T>);
 impl<T: frame_system::Config> cumulus_pallet_xcmp_queue::WeightInfo for SubstrateWeight<T> {
     // Storage: XcmpQueue QueueConfig (r:1 w:1)
     fn set_config_with_u32() -> Weight {
-        (6_354_000 as Weight)
+        (3_214_000 as Weight)
```
Comment (Garandor) [inline runtime/calamari/src/weights/cumulus_pallet_xcmp_queue.rs:57]: where do these 100% gains come from?

File: runtime/calamari/src/weights/frame_system.rs [hunk 4]
```diff
@@ -66,37 +67,37 @@ impl<T: frame_system::Config> frame_system::WeightInfo for SubstrateWeight<T> {
...
+            // Standard Error: 1_000
+            .saturating_add((494_000 as Weight).saturating_mul(i as Weight))
             .saturating_add(T::DbWeight::get().writes((1 as Weight).saturating_mul(i as Weight)))
     }
     // Storage: Skipped Metadata (r:0 w:0)
     fn kill_storage(i: u32, ) -> Weight {
-        (2_162_000 as Weight)
-            // Standard Error: 0
-            .saturating_add((346_000 as Weight).saturating_mul(i as Weight))
+        (1_114_000 as Weight)
```
Comment (Garandor) [inline runtime/calamari/src/weights/frame_system.rs:83]: also here

File: runtime/calamari/src/weights/pallet_treasury.rs [hunk 5]
```diff
@@ -44,100 +45,100 @@ use sp_std::marker::PhantomData;
 
 /// Weight functions needed for pallet_treasury.
 pub trait WeightInfo {
-	fn propose_spend() -> Weight;
-	fn reject_proposal() -> Weight;
-	fn approve_proposal(p: u32, ) -> Weight;
-	fn remove_approval() -> Weight;
-	fn on_initialize_proposals(p: u32, ) -> Weight;
+    fn propose_spend() -> Weight;
```
Comment (Garandor) [inline runtime/calamari/src/weights/pallet_treasury.rs:48]: why these formatting changes

File: runtime/calamari/src/weights/pallet_utility.rs [hunk 6]
```diff
@@ -44,60 +45,66 @@ use sp_std::marker::PhantomData;
 
 /// Weight functions needed for pallet_utility.
 pub trait WeightInfo {
-	fn batch(c: u32, ) -> Weight;
```
Comment (Garandor) [inline runtime/calamari/src/weights/pallet_utility.rs:47]: same here

File: runtime/dolphin/src/weights/frame_system.rs [hunk 7]
```diff
@@ -56,79 +57,79 @@ pub trait WeightInfo {
 pub struct SubstrateWeight<T>(PhantomData<T>);
 impl<T: frame_system::Config> frame_system::WeightInfo for SubstrateWeight<T> {
     fn remark(_b: u32, ) -> Weight {
-        (1_672_000 as Weight)
+        (6_824_000 as Weight)
```
Comment (Garandor) [inline runtime/dolphin/src/weights/frame_system.rs:60]: remark became 5x more expensive?

File: runtime/dolphin/src/weights/pallet_assets.rs [hunk 8]
```diff
@@ -74,13 +75,13 @@ pub struct SubstrateWeight<T>(PhantomData<T>);
 impl<T: frame_system::Config> pallet_assets::WeightInfo for SubstrateWeight<T> {
     // Storage: Assets Asset (r:1 w:1)
     fn create() -> Weight {
-        (20_622_000 as Weight)
+        (13_228_000 as Weight)
```
Comment (Garandor) [inline runtime/dolphin/src/weights/pallet_assets.rs:78]: creating an asset became 40% cheaper?

File: runtime/dolphin/src/weights/pallet_treasury.rs [hunk 9]
```diff
@@ -44,102 +45,100 @@ use sp_std::marker::PhantomData;
 
 /// Weight functions needed for pallet_treasury.
 pub trait WeightInfo {
-	fn propose_spend() -> Weight;
```
Comment (Garandor) [inline runtime/dolphin/src/weights/pallet_treasury.rs:47]: reformatted

File: runtime/dolphin/src/weights/pallet_utility.rs [hunk 10]
```diff
@@ -44,60 +45,66 @@ use sp_std::marker::PhantomData;
 
 /// Weight functions needed for pallet_utility.
 pub trait WeightInfo {
-	fn batch(c: u32, ) -> Weight;
-	fn as_derivative() -> Weight;
```
Comment (Garandor) [inline runtime/dolphin/src/weights/pallet_utility.rs:48]: reformatted

**Comment by Dengjianping** [comment]
@Garandor 
I don't think there's formatting issue, because all CI jobs have passed.
And I did `taplo fmt/lint` locally, every thing looks fine.

File: runtime/calamari/src/weights/cumulus_pallet_xcmp_queue.rs (hunk 3 above)
Comment (Dengjianping) [reply runtime/calamari/src/weights/cumulus_pallet_xcmp_queue.rs:57]: Actually it's the first time we benchmark this pallet in calamari.
We didn't benchmarked it in v3.2.0. And all extrinsics from this pallet are only available for council/root. No risk at all.

**Comment by Dengjianping** [comment]
Not sure there's chaos while benchmarking, caused dramatic change in these weights.
I will trigger benchmark test for dolphin again to see whether any big changes in weights.

**Comment by Garandor** [comment]
> @Garandor I don't think there's formatting issue, because all CI jobs have passed. And I did `taplo fmt/lint` locally, every thing looks fine.

The old weights were checked in and all jobs were passing then too.
Both can't be right.

Sounds to me like the CI job is failing to detect formatting issues reliably after all

In order to avoid blocking this release on formatting, let's create a followup

File: runtime/dolphin/src/weights/pallet_assets.rs (hunk 8 above)
Comment (Garandor) [reply runtime/dolphin/src/weights/pallet_assets.rs:78]: For some reason, the v3.2.0 release increased this to 20M.
It was ~15M before.

Dunno why chaos exists @Dengjianping since this weight updating process is well organized, but at least the new value is not an outlier.
I can't remember us ever talking about this in v3.2.0, so the manual detection of benchmarking changes is not reliable it seems if big changes like this stay under the radar.

I'd want a script/tool that compares the weight diffs for changes larger than ~10% as we were discussing back in the past, as scanning 50+ trivial diff weights files on every release seems not to be a job humans can or want to do.

cc @ghzlatarev @bhgomes

File: runtime/dolphin/src/weights/frame_system.rs (hunk 7 above)
Comment (Garandor) [reply runtime/dolphin/src/weights/frame_system.rs:60]: in v3.2.0 remark went f

```
-       fn remark(b: u32, ) -> Weight {
-               (8_427_000 as Weight)
-                       // Standard Error: 0
-                       .saturating_add((1_000 as Weight).saturating_mul(b as Weight))
+       fn remark(_b: u32, ) -> Weight {
+               (1_672_000 as Weight)
```
now it's back up. This large change went under the radar again

File: runtime/calamari/src/weights/cumulus_pallet_xcmp_queue.rs (hunk 3 above)
Comment (Dengjianping) [reply runtime/calamari/src/weights/cumulus_pallet_xcmp_queue.rs:57]: I re-benchmarked dolphin, the result looks the same:
```rust
fn set_config_with_u32() -> Weight {
        (3_294_000 as Weight)
            .saturating_add(T::DbWeight::get().reads(1 as Weight))
            .saturating_add(T::DbWeight::get().writes(1 as Weight))
    }
```

https://github.com/Manta-Network/Manta/actions/runs/2696127661

File: runtime/dolphin/src/weights/frame_system.rs (hunk 7 above)
Comment (Dengjianping) [reply runtime/dolphin/src/weights/frame_system.rs:60]: From latest result:
```rust
fn remark(_b: u32, ) -> Weight {
        (5_378_000 as Weight)
    }
```

File: runtime/dolphin/src/weights/pallet_assets.rs (hunk 8 above)
Comment (Dengjianping) [reply runtime/dolphin/src/weights/pallet_assets.rs:78]: From latest result:
```rust
fn create() -> Weight {
        (13_450_000 as Weight)
            .saturating_add(T::DbWeight::get().reads(1 as Weight))
            .saturating_add(T::DbWeight::get().writes(1 as Weight))
    }
```

Comment (ghzlatarev) [reply runtime/dolphin/src/weights/pallet_assets.rs:78]: We haven't benchmarked until now so now is the time to discuss these.
This is the existing issue https://github.com/Manta-Network/Manta/issues/399

**Review by ghzlatarev — APPROVED** [review]

**Review by Dengjianping — APPROVED** [review]
